src/storage/heap_file.c

```diff
@@ -25090,20 +25093,102 @@ heap_get_visible_version_with_repl_desync (THREAD_ENTRY * thread_p, HEAP_GET_CON
 	  break;
 	}
       assert (is_passive_transaction_server ());
+      assert (context->scan_cache != nullptr);
+
       // Handle the page desynchronization error
-      if (context->home_page_watcher.pgptr)
-	{
-	  /* Unfix home page. */
-	  pgbuf_ordered_unfix (thread_p, &context->home_page_watcher);
-	}
+      //
 
-      if (context->fwd_page_watcher.pgptr != NULL)
+      // clean - saving state if required
+      //
+      heap_clean_get_context (thread_p, context);
```

Comment:
I thought that you want to refix not only the home page and the forward page, but also a page saved in `context->scan_cache->page_watcher.pgptr`. However, it seems that `heap_clean_get_context` makes `page_watcher.pgptr` point to the home page so it ends up just refixing only home and forward pages just as before. Do I get it wrong?

Reply:
>However, it seems that heap_clean_get_context makes page_watcher.pgptr point to the home page 

in heap_clean_get_context()
```
25485   if (context->scan_cache != NULL && context->scan_cache->cache_last_fix_page
25486       && context->home_page_watcher.pgptr != NULL)
25487     {
25488       /* Save home page (or NULL if it had to be unfixed) to scan_cache. */
25489       pgbuf_replace_watcher (thread_p, &context->home_page_watcher, &context->scan_cache->page_watcher);
25490       assert (context->home_page_watcher.pgptr == NULL);
25491     }
```

Reply:
Right after `ER_PAGE_AHEAD_OF_REPLICATION` error is detected the intended logic is as follows:
- step `heap_clean_get_context`:
  - if there is a home page watcher, it is saved in `scan_cache` watcher
  - home and forward pages are cleaned (well, at this point only the forward pages might have something in the watcher)
- step `pgbuf_ordered_unfix_and_save_for_refix`:
  - if there was something saved in the `scan_cache` page wather, it is unfixed and the coordinates for refix are saved (latch mode and vpid)
- step: wait for replication
- step: if `restore_scan_cache_page_watcher`
  - if there was a `scan_cache` watcher active it is re-fixed with `OLD_PAGE_IF_IN_BUFFER_OR_IN_TRANSIT`
- step: `heap_init_get_context`
  -  will restore the `scan_cache` watcher (if present) to the home page wather

The purpose is to leave the state as it was as if there was no wait for replication.
With the caveat explained that the page might not be there anymore.

However, the forward page is purposefully not re-fixed, only the home will be re-fixed (via the `scan_cache` route as described above).

In that case we leave it up to the next call to `heap_get_visible_version_internal` to deal with such state.

Reply:
I understand every step you explain, but I don't think it still addresses my question.
> It was the case that the page held by scan_cache::page_watcher was not being released, leading to a dead-latch between client transaction thread and replication thread in the page buffer layer.

The purpose of this PR is to release a page that is possibly fixed in scan_cache::page_watcher. Do I get it right? Then, 
> step heap_clean_get_context:
if there is a home page watcher, it is saved in scan_cache watcher
home and forward pages are cleaned (well, at this point only the forward pages might have something in the watcher)

In the above,  scan_cache::page_watcher is overwritten with the home page and the forward page is unfixed in heap_clean_get_context(), and the home page (in scan_cache::page_watcher) is unfixed in the caller. After the home page is refixed after waiting and the forward page will be refixed in the next `heap_get_visible_version_internal` call. 
So, it seems to end up just refixing the home and forward pages just like before. That is my question. Sorry if I misunderstand something.

Reply:
Additional question: 
As far as I know, a thread also possibly keeps some other different pages fixed than the three pages above. Don't they matter for the issue case?

Reply:
The purpose is to:
- unfix all pages that are left fixed by the call to `heap_get_visible_version_internal`
- wait for replication
- refix only what is saved in `scan_cache` based on the idea that it is a "cache" and is needed to re-attempt to retrieve the data
- looking deeper at the call path `heap_get_visible_version_internal` -> `heap_prepare_get_context` -> `heap_prepare_object_page(.. &context->home_page_watcher ..)` I can see that the logic benefits from having the home page fixed (as restored from `scan_cache` by the last call to `heap_init_get_context`).

I still have a feeling that I don't manage to answer your question.
If you think the patch introduces some inconsistency or does "too much" or "too little" just name it. :)

> As far as I know, a thread also possibly keeps some other different pages fixed than the three pages above. Don't they matter for the issue case?

What other pages? Where?
This patch does address one found case of "keeps other different pages fixed".

Reply:
>unfix all pages that are left fixed by the call to heap_get_visible_version_internal 

The home and forward pages were unfixed also in the previous code, right? And you said we missed a page to unfix before waiting. It must be the page stored in `scan_cache::page_watcher`. Do I get it right? 
In the changed code, the way I see it, the `scan_cache::page_watcher` could be overwritten by the home page in `heap_clean_get_context()`, Then, we end up losing track of the page store in `scan_cache::page_watcher` previously and leaving it unfixed. Do I make it clear? 

in heap_clean_get_context():
```
25485   if (context->scan_cache != NULL && context->scan_cache->cache_last_fix_page
25486       && context->home_page_watcher.pgptr != NULL)
25487     {
25488       /* Save home page (or NULL if it had to be unfixed) to scan_cache. */
25489       pgbuf_replace_watcher (thread_p, &context->home_page_watcher, &context->scan_cache->page_watcher);
25490       assert (context->home_page_watcher.pgptr == NULL);
25491     }
```

P.S. Looking into `pgbuf_replace_watcher()` again, it expects `context->scan_cache->page_watcher` to point to nothing. Then, I guess that pgbuf_replace_watcher() is only called when there is no pgptr in `context->scan_cache->page_watcher`. And I think the above code snippet is not executed for the issue case and that `heap_clean_get_context()` is just another way to unfix the home and forward page. If it's just a short way to unfix them, I'm OK. I was worried that we possibly lose the `context->scan_cache->page_watcher`.

> What other pages? Where?
This patch does address one found case of "keeps other different pages fixed".

Generally speaking, a thread can fix many different pages and I guessed that there may be some pages fixed other than the three pages described. For example, before coming into `heap_get_visible_version_with_repl_desync()`, a heap header or the previous heap page during a heap scan could be fixed (I didn't see the case, I just thought about it generally). But if no pages are guaranteed to be fixed, just forget it. It would be good if there is an assertion that we can check if no pages are fixed by the thread, though.

Reply:
@wrlawodms : Indeed, the same transaction can have other pages already fixed from outer scans.
I expect the replication thread, wants to fix a bunch of pages belonging to the same file as part of an "atomic" replication operation (it will never want to fix pages from different heap files).
So, I think the problem can arise only when the outer scan is on the same heap file as the inner scan (as hint to reproduce the issue).
Regardless to repro, a possible solution could be to unfix all ordered fixed pages (all pages having watchers in the current transaction) and then refix all of them.

Reply:
@wrlawodms 

> The home and forward pages were unfixed also in the previous code, right?

Yes

> And you said we missed a page to unfix before waiting. It must be the page stored in scan_cache::page_watcher. Do I get it right?

Yes

> In the changed code, the way I see it, the scan_cache::page_watcher could be overwritten by the home page in heap_clean_get_context(), Then, we end up losing track of the page stored in scan_cache::page_watcher previously and leaving it unfixed. Do I make it clear?

Yes, the question is clear. But we cannot lose track of whatever was previously fixed in `scan_cache->page_watcher` because, as you also point out, the `scan_cache->page_watcher` is asserted empty in `pgbuf_replace_watcher`. However, this code is safe only in debug mode while in release mode, I suppose an overwrite into an non-empty `scan_cache->page_watcher` could happen.
On the other hand, this scenario could have happened before as well. So wherever this inconsistency might have happened it shoul have already happened (either in tests or in production). The newly introduced code does not introduce anything new in this aspect.

Reply:
@wrlawodms @razvanradulescu 

I looked further into the matter and I see two aspects of the problem which I can summarize as follows:

> If a PTS client transaction/thread must wait the replication thread to advance, it must also "make way" so the replication thread can advance. "Make way" means: unfix ALL pages that the transaction holds fixed before waiting on the replication thread to advance.

I looked at the code and I see one clear potential issue:
- on the call path `heap_next_internal` -> `heap_scan_get_visible_version` -> `heap_get_visible_version_with_repl_desync`
- in function `heap_next_internal` there are some other watchers - `curr_page_watcher` and `old_page_watcher` - which I don't think are guaranteed to point to nothing
- thus, it is possible that these are holding other pages fixed while the function `heap_get_visible_version_with_repl_desync` is performing a wait on replication thread
- thus, the potential for dead-latches is still there
- for this problem I will add another - separate - issue to implement a fix; at the very least we can add some checks with some `err_set` statements and debug asserts that will at least signal the problem to us;
- in short we'll deal with this in a separate issue

I also can imagine another issue but I know no clear scenario if this can be reproduced:
- is it possible in a client transaction to have outer and inner scans and as @razvanradulescu sais:

> So, I think the problem can arise only when the outer scan is on the same heap file as the inner scan (as hint to reproduce the issue).

- maybe this issue is in fact a more general one of the previous particular description (the one above starting from `heap_next_internal` function)
- I will also describe this and we'll take this in the separate issue

Reply:
Now everything is clear to me. Thank you for taking it seriously even without a concrete example.